File: tutorials/ParallelBenchmarkCases/B_O3_06/usr_rates.f
!vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv!
!                                                                      !
!  Module name: USR_RATES                                              !
!                                                                      !
!  Purpose:                                                            !
!                                                                      !
!  Author: J.Musser                                   Date: 10-Oct-12  !
!                                                                      !
!  Comments:                                                           !
!                                                                      !
!^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^!
      SUBROUTINE USR_RATES(IJK, RATES)

      USE param
      USE param1
      USE parallel
      USE fldvar
      USE rxns
      USE energy
      USE geometry
      USE run
      USE indices
      USE physprop
      USE constant
      USE funits
      USE compar
      USE sendrecv

      USE toleranc
      USE usr

      IMPLICIT NONE


      INTEGER, INTENT(IN) :: IJK

      DOUBLE PRECISION, DIMENSION(NO_OF_RXNS), INTENT(OUT) :: RATES

!-----------------------------------------------
      INCLUDE 'species.inc'

      INCLUDE 'ep_s1.inc'
      INCLUDE 'fun_avg1.inc'

      INCLUDE 'function.inc'

      INCLUDE 'ep_s2.inc'
      INCLUDE 'fun_avg2.inc'

      INCLUDE 'usrnlst.inc'

! Reaction specific variables:
!`````````````````````````````````````````````````````````````````````//
      DOUBLE PRECISION c_O3   ! Ozone concentration mol/cm^3

! PhseCng:  O3 --> 1.5O2         (mol/cm^3.s)
!---------------------------------------------------------------------//
      c_O3 = (RO_g(IJK)*X_g(IJK,O3)/MW_g(O3))

      RATES(Ozone_Decomp) = (ONE - EP_g(IJK)) * C(1) * c_O3

      RETURN

      END SUBROUTINE USR_RATES

File: tutorials/reactor1a/usr_rates.py
# !vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv!
# !                                                                      !
# !  Module name: USR_RATES                                              !
# !                                                                      !
# !  Purpose:                                                            !
# !                                                                      !
# !  Author: J.Musser                                   Date: 10-Oct-12  !
# !                                                                      !
# !  Comments:                                                           !
# !                                                                      !
# !^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^!

from mfix import constant as CONSTANT
from mfix import fldvar as FLDVAR
from mfix import physprop as PHYSPROP

def usr_rates(IJK, RATES):

    for line in open('species.inc').readlines():
        if '::' in line:
            exec(line.split('::')[1].strip())

    # ! Reaction specific variables:
    # !`````````````````````````````````````````````````````````````````````//
    #       DOUBLE PRECISION c_A    ! Species A concentration mol/cm^3

    # ! AtoR:  A --> 3.0*R         (mol/cm^3.s)
    # !---------------------------------------------------------------------//

    c_A = FLDVAR.ro_g[IJK-1]*FLDVAR.x_g[IJK-1,A-1]/PHYSPROP.mw_g[A-1]

    RATES[AtoR-1] = FLDVAR.ep_g[IJK-1] * CONSTANT.c[1-1] * (c_A)**CONSTANT.c[2-1]
